File: src/main/java/ee/tenman/portfolio/alphavantage/AlphaVantageService.java
package ee.tenman.portfolio.alphavantage;

import com.google.gson.Gson;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.retry.annotation.Backoff;
import org.springframework.retry.annotation.Retryable;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Optional;

@Service
@Slf4j
public class AlphaVantageService {
	
	private static final Gson GSON = new Gson();
	
	@Resource
	private AlphaVantageClient client;
	
	@Retryable(backoff = @Backoff(delay = 1000))
	public AlphaVantageResponse getMonthlyTimeSeries(final String symbol) {
		return this.getTicker(symbol)
				.map(ticker -> {
					final AlphaVantageResponse timeSeriesMonthly = this.client.getMonthlyTimeSeries("TIME_SERIES_MONTHLY", ticker);
					log.info("Retrieved monthly ticker data: {}", GSON.toJson(timeSeriesMonthly));
					return timeSeriesMonthly;
				})
				.orElseThrow(() -> new RuntimeException("Error while fetching data from Alpha Vantage"));
	}
	
	@Retryable(backoff = @Backoff(delay = 1000))
	public Optional<String> getTicker(final String search) {
		final SearchResponse symbolSearch = this.client.getSearch("SYMBOL_SEARCH", search, String.valueOf(1));
		return Optional.ofNullable(symbolSearch)
				.map(SearchResponse::getBestMatches)
				.stream()
				.flatMap(Collection::stream)
				.findFirst()
				.map(SearchResponse.SearchData::getSymbol);
	}
	
}

File: src/main/kotlin/ee/tenman/portfolio/alphavantage/AlphaVantageService.kt
package ee.tenman.portfolio.alphavantage

import ee.tenman.portfolio.alphavantage.AlphaVantageResponse.AlphaVantageDailyPriceData
import ee.tenman.portfolio.configuration.JsonMapperFactory
import jakarta.annotation.Resource
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Service
class AlphaVantageService {
  private val log = LoggerFactory.getLogger(javaClass)

  companion object {
    private val DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd")
  }

  @Resource
  private lateinit var client: AlphaVantageClient

  fun getDailyTimeSeriesForLastWeek(symbol: String): Map<LocalDate, AlphaVantageDailyPriceData> {
    var adjustedSymbol = symbol
    if ("QDVE:GER:EUR" == symbol) {
      adjustedSymbol = "QDVE.DEX"
    }

    val ticker =
      getTicker(adjustedSymbol) ?: run {
        log.error("Failed to get ticker for symbol: $adjustedSymbol")
        return emptyMap()
      }

    return try {
      val timeSeriesDaily = client.getTimeSeries("TIME_SERIES_DAILY", ticker)
      val json = JsonMapperFactory.instance.writeValueAsString(timeSeriesDaily)
      log.info("Retrieved daily ticker data for $ticker: ${JsonMapperFactory.truncateJson(json)}")

      timeSeriesDaily.dailyTimeSeries
        ?.asSequence()
        ?.associate { (dateString, data) ->
          LocalDate.parse(dateString, DATE_FORMATTER) to data
        }?.toMap() ?: emptyMap()
    } catch (e: Exception) {
      log.error("Error fetching daily data from Alpha Vantage for ticker $ticker", e)
      emptyMap()
    }
  }

  fun getTicker(search: String): String? {
    val symbolSearch = client.getSearch("SYMBOL_SEARCH", search, "1")
    return symbolSearch.bestMatches?.firstOrNull()?.symbol
  }
}
